File: ProcessaTexto/src/main/kotlin/br/com/fenix/processatexto/model/entities/firebase/Vocabulario.java
package br.com.fenix.processatexto.model.entities.firebase;

import com.google.gson.annotations.Expose;

import java.util.HashMap;
import java.util.Objects;
import java.util.UUID;

public class Vocabulario {
    protected UUID id;
    @Expose
    protected String vocabulario;
    protected String formaBasica;
    @Expose
    protected String leitura;
    protected String leituraNovel;
    @Expose
    protected String ingles;
    @Expose
    protected String portugues;
    public String sincronizacao;

    public UUID getId() {
        return id;
    }

    public void setId(UUID id) {
        this.id = id;
    }

    public String getVocabulario() {
        return vocabulario;
    }

    public void setVocabulario(String vocabulario) {
        this.vocabulario = vocabulario;
    }

    public String getFormaBasica() {
        return formaBasica;
    }

    public void setFormaBasica(String formaBasica) {
        this.formaBasica = formaBasica;
    }

    public String getLeitura() {
        return leitura;
    }

    public void setLeitura(String leitura) {
        this.leitura = leitura;
    }

    public String getLeituraNovel() {
        return leituraNovel;
    }

    public void setLeituraNovel(String leituraNovel) {
        this.leituraNovel = leituraNovel;
    }

    public String getIngles() {
        return ingles;
    }

    public void setIngles(String ingles) {
        this.ingles = ingles;
    }

    public String getPortugues() {
        return portugues;
    }

    public void setPortugues(String portugues) {
        this.portugues = portugues;
    }


    public Vocabulario(br.com.fenix.processatexto.model.entities.processatexto.Vocabulario vocabulario) {
        this.id = vocabulario.getId();
        this.vocabulario = vocabulario.getVocabulario();
        this.formaBasica = vocabulario.getFormaBasica();
        this.leitura = vocabulario.getLeitura();
        this.leituraNovel = vocabulario.getLeituraNovel();
        this.portugues = vocabulario.getPortugues();
        this.ingles = vocabulario.getIngles();
        this.sincronizacao = vocabulario.getSincronizacao();
    }

    public Vocabulario(UUID id, String vocabulario, String formaBasica, String leitura, String leituraNovel, String ingles, String portugues) {
        this.id = id;
        this.vocabulario = vocabulario;
        this.formaBasica = formaBasica;
        this.leitura = leitura;
        this.leituraNovel = leituraNovel;
        this.portugues = portugues;
        this.ingles = ingles;
    }

    public static br.com.fenix.processatexto.model.entities.processatexto.Vocabulario toVocabulario(String id, HashMap<String, ?> obj) {
        return new br.com.fenix.processatexto.model.entities.processatexto.Vocabulario(UUID.fromString(id), (String) obj.get("vocabulario"), (String) obj.get("formaBasica"),
                (String) obj.get("leitura"), (String) obj.get("leituraNovel"), (String) obj.get("ingles"), (String) obj.get("portugues"));
    }

    @Override
    public String toString() {
        return vocabulario;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Vocabulario that = (Vocabulario) o;
        return Objects.equals(id, that.id) && Objects.equals(vocabulario, that.vocabulario);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, vocabulario);
    }
}

File: ProcessaTexto/src/main/kotlin/br/com/fenix/processatexto/model/entities/processatexto/Vocabulario.kt
package br.com.fenix.processatexto.model.entities.processatexto

import br.com.fenix.processatexto.model.entities.EntityBase
import com.google.gson.annotations.Expose
import jakarta.persistence.Column
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import org.hibernate.annotations.JdbcTypeCode
import org.hibernate.type.SqlTypes
import java.util.*


@jakarta.persistence.Entity
open class Vocabulario(
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @JdbcTypeCode(SqlTypes.VARCHAR)
    @Column(name = "ID", nullable = false, unique = true, length = 36)
    private var id: UUID? = null,
    @Column(length = 250, nullable = false, unique = true)
    @Expose open var vocabulario: String = "",
    @Column(name = "forma_basica", length = 250, nullable = true)
    open var formaBasica: String = "",
    @Column(length = 250, nullable = true)
    @Expose open var leitura: String = "",
    @Column(name = "leitura_novel", length = 250, nullable = true)
    open var leituraNovel: String = "",
    @Column(nullable = true)
    @Expose open var ingles: String = "",
    @Column(nullable = true)
    @Expose open var portugues: String = ""
) : EntityBase<UUID?, Vocabulario>() {

    @Transient
    var sincronizacao: String = ""

    constructor(vocabulario: String) : this(null, vocabulario)

    constructor(vocabulario: String, portugues: String) : this(null, vocabulario) {
        this.portugues = portugues
    }

    constructor(id: UUID?, vocabulario: String, portugues: String) : this(id, vocabulario) {
        this.portugues = portugues
    }

    constructor(vocabulario: String, formaBasica: String, leitura: String, leituraNovel: String) : this(null, vocabulario, formaBasica, leitura, leituraNovel)

    constructor(id: String, obj: HashMap<String, String>) : this(UUID.fromString(id)) {
        vocabulario = obj["vocabulario"]!!
        formaBasica = obj["formaBasica"]!!
        leitura = obj["leitura"]!!
        leituraNovel = obj["leituraNovel"]!!
        portugues = obj["portugues"]!!
        ingles = obj["ingles"]!!
    }

    @Override
    override fun toString(): String {
        return vocabulario
    }

    fun merge(vocab: Vocabulario) {
        vocabulario = vocab.vocabulario
        formaBasica = vocab.formaBasica
        leitura = vocab.leitura
        leituraNovel = vocab.leituraNovel
        portugues = vocab.portugues
        ingles = vocab.ingles
    }

    override fun getId(): UUID? = id

    override fun create(id: UUID?): Vocabulario = Vocabulario(id)

    open fun setId(id: UUID?) {
        this.id = id
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Vocabulario

        if (id != other.id) return false
        if (vocabulario != other.vocabulario) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id?.hashCode() ?: 0
        result = 31 * result + vocabulario.hashCode()
        return result
    }

}